### PR TITLE
add tests for http engine

### PR DESCRIPTION
**Pull request by flentini**

### CODE REVIEW

File: test/core/unit/engine_http.test.js [hunk 1]
```diff
@@ -112,26 +118,210 @@ test('HTTP virtual user', function (t) {
...
+  const engine = new HttpEngine(script);
+  const ee = new EventEmitter();
+  const runScenario = engine.createScenario(script.scenarios[0], ee);
+
+  runScenario({ vars: {} }, function userDone(err, context) {
+    if (err) {
+      t.fail();
+    }
+
+    t.equal(context._successCount, 1, 'it should handle compressed responses');
```
Comment (hassy) [inline test/core/unit/engine_http.test.js:271]: * would it make sense to roll the check from the previous test (whether `accept-encoding` header is set) into here to reduce the amount of test code?
* `_successCount` does not seem to be used anywhere else in the codebase (even though it seems to be set by all engines) - would suggest not relying on it here and getting rid of it
* we're not actually testing that the response is decompressed by the client as expected here? (even checking `_successCount` does not tell us that we have the decompressed response)

File: test/core/unit/engine_http.test.js [hunk 2]
```diff
@@ -141,11 +331,11 @@ test('url and uri parameters', function (t) {
     config: {
       target: 'http://localhost:8888',
       processor: {
-        rewriteUrl: function (req, context, ee, next) {
+        rewriteUrl: function (req, _context, _ee, next) {
```
Comment (hassy) [inline test/core/unit/engine_http.test.js:334]: should we add prefixing with `_`  to unused variables to lint rules?

File: test/core/unit/engine_http.test.js [hunk 3]
```diff
@@ -247,21 +435,152 @@ test('hooks - afterResponse', (t) => {
...
+test('hooks - beforeScenario', (t) => {
+  const endpoint = '/products';
+  const script = {
+    config: {
+      target: 'http://localhost:8888',
+      processor: {
+        setEndpoint: function (context, ee, next) {
+          context.vars.endpoint = endpoint;
+
+          t.ok(
```
Comment (hassy) [inline test/core/unit/engine_http.test.js:453]: Is the goal here to check the signature of hook functions?

If so, should we have some assertions on other parameters too?

File: test/core/unit/engine_http.test.js [hunk 4]
```diff
@@ -247,21 +435,152 @@ test('hooks - afterResponse', (t) => {
...
+
+test('hooks - afterScenario', (t) => {
+  const endpoint = '/products';
+  const productsCount = 123;
+  const script = {
+    config: {
+      target: 'http://localhost:8888',
+      processor: {
+        checkProductsCount: function (context, ee, next) {
+          t.ok(
```
Comment (hassy) [inline test/core/unit/engine_http.test.js:511]: as above - if the reason for this check is to check the interface of hook functions then we should check other args + maybe do it in one place to DRY?

File: test/core/unit/engine_http.test.js [hunk 5]
```diff
@@ -247,21 +435,152 @@ test('hooks - afterResponse', (t) => {
...
+
+          t.equal(
+            context.vars.count,
+            productsCount,
+            'it can access variables set by the scenario'
+          );
+
+          t.ok(
+            context.vars.date === undefined,
+            'it cannot access variables set by other scenarios'
```
Comment (hassy) [inline test/core/unit/engine_http.test.js:524]: Nice idea - but it seems like the second scenario is never run, so the assertion can never fail?

File: test/core/test_capture.js [hunk 6]
```diff
@@ -1,27 +1,71 @@
...
 
 let xmlCapture = null;
 try {
   xmlCapture = require('artillery-xml-capture');
 } catch (e) {}
 
+beforeEach(() => nock.cleanAll());
+
 test('Capture - headers', (t) => {
-  const fn = path.resolve(__dirname, './scripts/captures-header.json');
```
Comment (hassy) [inline test/core/test_capture.js:17]: The file does not seem to be used anywhere else? should probably delete it altogether if so

File: test/core/test_capture.js [hunk 7]
```diff
@@ -30,88 +74,178 @@ test('Capture - headers', (t) => {
...
-        t.assert(cond, 'There should be a 200 for every 201');
-        if (!cond) {
-          console.log(report);
-          console.log('200: %s; 201: %s', c200, c201);
-        }
-        ee.stop().then(() => {
-          t.end();
-        });
-      });
+test('Capture - selector', (t) => {
```
Comment (hassy) [inline test/core/test_capture.js:77]: Nice 👍

File: test/core/test_capture.js [hunk 8]
```diff
@@ -30,88 +74,178 @@ test('Capture - headers', (t) => {
   });
 });
 
-test('Capture - JSON', (t) => {
-  const fn = path.resolve(__dirname, './scripts/captures.json');
```
Comment (hassy) [inline test/core/test_capture.js:34]: Should remove this and all other JSON scripts that have been inlined into tests if they're not used anywhere else.

File: test/core/test_capture.js [hunk 9]
```diff
@@ -122,20 +256,96 @@ test('Capture before test - JSON', (t) => {
...
+              url: '/pets',
+              json: { name: '{{ name }}', species: '{{ species }}' },
+              capture: [
+                {
+                  json: '{{{ jsonPathExpr }}}',
+                  as: 'pet[id]'
+                },
+                {
+                  json: '$.doesnotexist',
+                  transform: 'this.doesnotexist.toUpperCase()',
```
Comment (hassy) [inline test/core/test_capture.js:290]: `transform` can be removed (here and everywhere else)

File: test/core/test_capture.js [hunk 10]
```diff
@@ -122,20 +256,96 @@ test('Capture before test - JSON', (t) => {
   });
 });
 
-test('Capture after test - JSON', (t) => {
-  const fn = path.resolve(__dirname, './scripts/after_test.json');
-  const script = require(fn);
-  runner(script).then(function (ee) {
+test('Capture and save to attribute of an Object in context.vars - JSON', (t) => {
```
Comment (hassy) [inline test/core/test_capture.js:259]: Might be staring at right at it and not seeing it but what's the difference between this test and the previous one? :D

Comment (flentini) [reply test/core/test_capture.js:259]: the first test captures a variable (`id`), the second captures a variable (`id`) as an object (`pet`) property

Comment (hassy) [reply test/core/test_capture.js:258]: so the second one makes the first one unnecessary then?

**Pull request merged**